### PR TITLE
Extend getBasis

### PR DESCRIPTION
**Pull request by akva2**
As discussed, this extends the getBasis method with the option to obtain additional bases
- geometry basis
- projection basis
- alternative projection basis
- refinement basis
- integration basis

I have added an enumeration to use for readability. 

I have also unified all code such that all bases are stored in their respective base classes (ASMstruct or ASMLRSpline).
This is not just for having consistent code, but I also have changes, currently only local, to add support for alternative projection bases in non-mixed ASMs which I need for adaptive equal-order simulators in Stokes. These additions
can be greatly simplified now, basically I can now move and reuse the code from the mixed ASMs. Only thing missing is that I would really like to use shared ptrs to simplify the pointer juggling in the structured ASMs, but I know you're not too fond of these so I not done that for now.

There are a few loose ends but this mostly sets the stage for adding support for a geometry basis separate from finite element bases.

### CODE REVIEW

**Comment by akva2** [comment]
jenkins build this with downstreams please

**Comment by akva2** [comment]
jenkins build this with downstreams please

**Comment by akva2** [comment]
jenkins build this with downstreams ifem-stokes=231 please

**Comment by akva2** [comment]
jenkins build this with downstreams ifem-stokes=231 please

**Review by kmokstad — COMMENTED** [review]
I started the review this on, until I realize it is still progressing (yes, you marked it as Draft also). So I stop here for now and let you finish it (with a rebase also).

However, consider if you agree on my suggestion in #603 first.

**Comment by akva2** [comment]
jenkins build this with downstreams please

**Comment by akva2** [comment]
jenkins build this with downstreams ifem-stokes=231 please

File: src/ASM/LR/ASMu2DC1.C [hunk 1]
```diff
@@ -18,7 +18,8 @@
 
 bool ASMu2DC1::generateFEMTopology ()
 {
-  if (!(geomB = this->createLRfromTensor()))
+    geomB = std::shared_ptr<LR::LRSpline>(this->createLRfromTensor());
+    if (!geomB)
     return false;
```
[the remark this answers is not in the document]
Comment (akva2) [reply src/ASM/LR/ASMu2DC1.C:23]: fixed.

File: src/ASM/ASMenums.h [hunk 2]
```diff
@@ -46,6 +46,12 @@ namespace ASM //! Assembly scope
     FULL_CACHE  //!< Cache basis function values up front
   };
 
+  //! \brief Enumeration of different basis types.
+  //! \details Entries should have negative values
+  enum BasisType {
+      GEOMETRY_BASIS   = 0,  //!< Geometry basis
+      PROJECTION_BASIS = -1, //!< Projection basis
+  };
```
[the remark this answers is not in the document]
Comment (akva2) [reply src/ASM/ASMenums.h:57]: fixed.

File: src/ASM/ASMs2D.C [hunk 3]
```diff
@@ -102,6 +102,17 @@ Go::SplineCurve* ASMs2D::getBoundary (int dir, int)
...
 
 
+Go::SplineSurface* ASMs2D::getBasis (int basis) const
+{
+  if (basis == ASM::GEOMETRY_BASIS)
+    return static_cast<Go::SplineSurface*>(geomB);
+  else if (basis == ASM::PROJECTION_BASIS)
+    return static_cast<Go::SplineSurface*>(projB);
+  else
+    return surf;
```
[the remark this answers is not in the document]
Comment (akva2) [reply src/ASM/ASMs2D.C]: yeah, using switch now.

**Comment by akva2** [comment]
jenkins build this with downstreams ifem-stokes=231 please

**Comment by akva2** [comment]
jenkins build this with downstreams ifem-stokes=231 please

**Comment by akva2** [comment]
jenkins build this with downstreams ifem-stokes=231 please

**Pull request closed without merging**